121 selecting multiple buildings

Heldereim opened the pull request:
It should now be possible to select multiple buildings at once. This is done by pressing the left mouse button and dragging the mouse thereby selecting an area on the playing field.

Selecting a single building should work the same as before. Selecting multiple buildings means that the information panel in the building info menu gets replaced by a list of how many buildings and of which type were selected.

It is possible to sell items from the selected buildings and set the production mode (selling or storing) of all selected buildings at the same time.

Heldereim committed:
Began work with box selection

Added a node for the box selection as well as an associated script. Also added this node to the game scene.

Heldereim committed:
Building selector moved and updated

Moved building selector script. Added a signal emitted when a selection is made.

Heldereim committed:
Selector updates.

Made the selector logic only run during Idle state. Started implementing the logic for showing multiple selected buildings in the building info panel.

Heldereim committed:
Renamed vboxcontainer

Heldereim committed:
Stuff done

Heldereim committed:
More work for multi select

Multi selecting buildings should be finished. Might be unfound bugs left.

Heldereim committed:
Warehouse clickable removed

Heldereim committed:
Merge branch '107-improve-smoke-and-emissions-from-buildings' into 121-selecting-multiple-buildings

Heldereim committed:
Bugfixes

Fixed a bug where biomass and other byproducts would not update properly in the building info panels storage tab. Also fixed an issue where the value of sold resources didn't update correctly.

linuskar commented on scripts/UI/BuildingInfo.gd:
```diff
@@ -275,3 +307,38 @@ func _sell_chosen_resources() -> void:
+
+func clean_buildings_list(buildings: Array[Building]) -> Array[Building]:
+	var cleaned_list: Dictionary[Building, bool] = {}
+	for building in buildings:
```
Why not just use building? Why have get_building?

Heldereim replied:
The reason this was implemented was so that landfills that automatically expanded could still be selected and if such was the case they would return the main node rather than the expanded instances.

linuskar commented on scripts/buildings/building.gd:
```diff
@@ -72,3 +73,6 @@ func _on_research_completed(id: String) -> void:
 func apply_research_upgrade() -> void:
 	pass
+
+func get_building() -> Building:
```
Why have get_building?

Heldereim replied:
This specific function isn't necessary for most buildings but specificially for expanding landfill this is needed since in those cases the 'building' isn't the node itself but rather the main node, that is the original landfill (whether that is biomass landfill or steelscrap landfill)

linuskar commented on scripts/buildings/storage_building.gd:
```diff
@@ -51,8 +54,8 @@ func init_storage_building() -> void:
 	output_storage.set(resource_type, resource_quantity - amount)
-	
 	input_storage.set(resource_type, resource_quantity - amount)
+	resources_changed.emit(resource_type, -amount)
```
Shouldn't the difference here for resources_changed signal be resource_quantity - amount?

linuskar commented on scripts/UI/BuildingInfo.gd:
```diff
@@ -12,33 +12,37 @@ class_name BuildingInfo extends UIMenu
-		for resource in storage_connections.keys():
-			var stored_resources: int = current_building.output_storage.get(resource)
-			storage_connections.get(resource).resource_held = stored_resources
-			if not Enums.is_byproduct(resource):
```
What does the update_storage function do now?

linuskar reviewed (changes requested):
- Can you update from main
- Unused clickable instance variable can be removed from biomass landfill
- In the error log fix so that RectangleShape2D size does not become negative
- Look at comments on code

Heldereim committed:
Removed unused functions

Removed update_storage and process functions from buildinginfo since they were no longer used

Heldereim committed:
Merge branch 'main' into 121-selecting-multiple-buildings

Heldereim committed:
Removed clickable instance variable

Heldereim committed:
Fixed a debug issue

linuskar reviewed (approved):
LGTM
